File: leetcode/medium/GasStation.py
from typing import List

class Solution:
    def canCompleteCircuit(self, gas: List[int], cost: List[int]) -> int:
        """
        Finds the starting gas station's index if you can travel around the circuit
        once in the clockwise direction, otherwise returns -1.

        This problem can be solved using a greedy approach in a single pass (O(n)).
        The logic relies on two key insights:

        1. If the total amount of gas available is less than the total cost to travel
           the entire circuit, it's impossible to complete the trip. We can check
           this first. `sum(gas) < sum(cost)`.

        2. If we start at station `A` and run out of gas trying to get to station `C`
           (meaning our tank becomes negative at some station `B` between `A` and `C`),
           then no station from `A` to `B` can be a valid starting point. This is
           because if we started at any station between `A` and `B`, we would have
           arrived at `B` with even less gas than when we started at `A`, and would
           still fail. Therefore, the next potential starting point must be `B+1`.

        Combining these, we iterate through the stations, keeping track of a running
        `tank` total. If the tank drops below zero, we know our current starting
        point is invalid, and we set the next station as the new potential start.
        Because of insight #1, if a solution exists, this process is guaranteed to find it.

        Args:
            gas: A list of integers where gas[i] is the amount of gas at station i.
            cost: A list of integers where cost[i] is the cost to travel from
                  station i to i+1.

        Returns:
            The index of the starting gas station, or -1 if no solution exists.
        """
        # Insight 1: Check if a solution is possible at all.
        if sum(gas) < sum(cost):
            return -1

        # Insight 2: Find the starting point using a greedy approach.
        tank = 0
        start_station = 0

        for i in range(len(gas)):
            # The net change in gas after visiting station `i` and traveling to `i+1`.
            tank += gas[i] - cost[i]

            # If the tank becomes negative, we can't reach station `i+1` from our
            # current `start_station`.
            if tank < 0:
                # Based on insight #2, none of the stations from `start_station`
                # to `i` can be the answer. So, the next potential start is `i+1`.
                start_station = i + 1
                # Reset the tank for the new journey starting from `i+1`.
                tank = 0
        
        # If the loop completes, the `start_station` found is the correct one.
        # This is guaranteed because we already know a solution exists (from insight #1).
        return start_station

File: leetcode/medium/GasStation.java
class Solution {
    /**
     * Determines the starting gas station index from which a car can complete a
     * full circular trip.
     *
     * The car has to travel from station 'i' to 'i+1'. The gas needed is cost[i].
     * If the total gas available is less than the total cost, no solution exists.
     * If a solution exists, it is guaranteed to be unique.
     *
     * @param gas An array where gas[i] is the amount of gas available at station i.
     * @param cost An array where cost[i] is the amount of gas needed to travel
     * from station i to station i+1 (or station 0 from the last station).
     * @return The starting index if a full circuit can be completed, or -1 otherwise.
     *
     * Time Complexity: O(n) - Single pass through the arrays.
     * Space Complexity: O(1) - Constant extra space used.
     */
    public int canCompleteCircuit(int[] gas, int[] cost) {
        // Essential check: An empty or null array means no circuit is possible.
        if (gas == null || gas.length == 0) {
            return -1;
        }

        // 1. Calculate the net change in fuel at each station: gas[i] - cost[i].
        //    Also, track the total gas and total cost to check for feasibility.
        int totalGas = 0;
        int totalCost = 0;
        int currentTank = 0; // Tracks the current net fuel balance in the tank for the *potential* circuit
        int startStation = 0; // The candidate starting station

        for (int i = 0; i < gas.length; i++) {
            totalGas += gas[i];
            totalCost += cost[i];

            // The net change in fuel after stopping at station 'i' and traveling to the next
            int netChange = gas[i] - cost[i];
            currentTank += netChange;

            // If the current tank balance drops below zero, it means the car cannot
            // reach station (i+1) starting from 'startStation'.
            // The new candidate must be the *next* station, (i+1), because any station
            // before or at 'i' cannot be the start of a valid circuit (as we've just proved).
            if (currentTank < 0) {
                // Reset the tank balance to 0, as we are effectively starting a new trip
                // from the next station.
                currentTank = 0;
                // The new potential starting point is the station immediately after the failure point
                startStation = i + 1;
            }
        }

        // 2. Feasibility check: If totalGas < totalCost, the trip is impossible.
        //    If totalGas >= totalCost, a solution is guaranteed to exist (unique).
        //    The unique solution will be the 'startStation' found in the single pass.
        return (totalGas < totalCost) ? -1 : startStation;
    }
}